openfe/protocols/openmm_utils/charge_generation.py

```diff
@@ -408,12 +414,106 @@ def assign_offmol_partial_charges(
 
-    # Call selected method to assign partial charges
-    CHARGE_METHODS[method.lower()]['charge_func'](
-        offmol=offmol_copy,
-        toolkit_registry=toolkits,
-        **CHARGE_METHODS[method.lower()]['charge_extra_kwargs'],
-    )  # type: ignore
+    # limit the number of threads used by SQM
+    # <https://github.com/openforcefield/openff-toolkit/issues/1831>
+    with temp_env({"OMP_NUM_THREADS": "2"}):
+        # Call selected method to assign partial charges
+        CHARGE_METHODS[method.lower()]['charge_func'](
+            offmol=offmol_copy,
+            toolkit_registry=toolkits,
+            **CHARGE_METHODS[method.lower()]['charge_extra_kwargs'],
+        )  # type: ignore
 
     # Copy partial charges back
     offmol.partial_charges = offmol_copy.partial_charges
+    return offmol
+
+
+def bulk_assign_partial_charges(
+    molecules: list[SmallMoleculeComponent],
+    overwrite: bool,
+    method: Literal['am1bcc', 'am1bccelf10', 'nagl', 'espaloma'],
+    toolkit_backend: Literal['ambertools', 'openeye', 'rdkit'],
+    generate_n_conformers: Optional[int],
+    nagl_model: Optional[str],
+    processors: int = 1,
+) -> list[SmallMoleculeComponent]:
+    """
+    Assign partial charges to a list of SmallMoleculeComponents using multiprocessing.
+
+    Parameters
+    ----------
+    molecules : list[gufe.SmallMoleculeComponent]
+      The list of molecules who should have partial charges assigned.
+    overwrite : bool
+      Whether or not to overwrite any existing non-zero partial charges.
+      Note that zeroed charges will always be overwritten.
+    method : Literal['am1bcc', 'am1bccelf10', 'nagl', 'espaloma']
+      Partial charge assignment method.
+      Supported methods include; am1bcc, am1bccelf10, nagl, and espaloma.
+    toolkit_backend : Literal['ambertools', 'openeye', 'rdkit']
+      OpenFF toolkit backend employed for charge generation.
+      Supported options:
+        * ``ambertools``: selects both the AmberTools and RDKit Toolkit Wrapper
+        * ``openeye``: selects the OpenEye toolkit Wrapper
+        * ``rdkit``: selects the RDKit toolkit Wrapper
+      Note that the ``rdkit`` backend cannot be used for `am1bcc` or
+      ``am1bccelf10`` partial charge methods.
+    generate_n_conformers : Optional[int]
+      Number of conformers to generate for partial charge generation.
+      If ``None`` (default), the input conformer will be used.
+      Values greater than 1 can only be used alongside ``am1bccelf10``.
+    nagl_model : Optional[str]
+      The NAGL model to use for charge assignment if method is ``nagl``.
+      If ``None``, the latest am1bcc NAGL charge model is used.
+    processors: int, default 1
+        The number of processors which should be used to generate the charges.
+
+    Raises
+    ------
+    ValueError
+      If the ``toolkit_backend`` is not supported by the selected ``method``.
+      If ``generate_n_conformers`` is ``None``, but the input ``offmol``
```

Comment:
This might actually be the case where we want to not inplace modify charges - what if things fails half way through a charge assignment loop, do we end up with have half our ligands charged and half our ligands uncharged?

Reply:
Might be worth testing - could be in a follow-up (for the sake of the CLI if this fails the whole thing fails, but the API is a bit different).